lib/ecl/ecl_file_kw.cpp

```diff
@@ -403,16 +403,16 @@ ecl_file_kw_type ** ecl_file_kw_fread_alloc_multiple( FILE * stream , int num) {
         buffer_offset += ECL_STRING8_LENGTH;
       }
 
-      kw_size = *((int *) &buffer[ buffer_offset ]);
+      memcpy(&kw_size, &buffer[ buffer_offset ], sizeof kw_size);
```

Comment:
why we need to copy here? it's used in `ecl_file_kw_alloc0` which only passes the value further on and not the reference

Reply:
Both versions do a copy. The problem with casting it instead of explicitly copying is undefined behaviour due to aligned issues. Some architectures don't support `mov`-ing data from misaligned addresses and will crash. This isn't the case for either x86 or any other remotely relevant architecture for us, but it's still UB.

`*((int *) &x)` roughly means "assume that x is an int", which will create `mov` instructions for copying the "int" `x` into a different memory location. Suppose that the architecture only supports integers that are aligned to 16 bytes (that is, their hex memory address _must_ end with 0). `*((int *) &x)` cannot guarantee this, and so it will crash if we first read 1 byte and then try to read an int from the buffer. However, `memcpy` is a simple byte for byte copy, and so will always work.

Both version generate the same assembly with GCC on x86. https://godbolt.org/z/XDwbYF

Reply:
valid point!